# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/BrowserChild.cpp
# Commit: b8ad0267e88a
# Full Hash: b8ad0267e88abff0a54a6f0f1b5de2ad3a018460
# Author: Hiroyuki Ikezoe <hikezoe.birchill@mozilla.com>
# Date: 2020-02-11 15:57:30
# Regressor Bug: 1599795
# File Overlap Count: 1
# Description:
#   Bug 1599795 - Make IntersectionObserver work in fission world. r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D61938
# ==============================================================================

diff -r 3985a60a75b3 -r b8ad0267e88a dom/ipc/BrowserChild.cpp
--- a/dom/ipc/BrowserChild.cpp	Tue Feb 11 03:35:56 2020 +0000
+++ b/dom/ipc/BrowserChild.cpp	Tue Feb 11 03:59:14 2020 +0000
@@ -3394,6 +3394,41 @@
   return Some(visibleRectLD);
 }
 
+Maybe<LayoutDeviceRect>
+BrowserChild::GetTopLevelViewportVisibleRectInSelfCoords() const {
+  if (mIsTopLevel) {
+    return Nothing();
+  }
+
+  if (!mChildToParentConversionMatrix) {
+    // There is a case that mChildToParentConversionMatrix hasn't been delivered
+    // since no APZ stuff has happened. In the case we can use mVisibleRect
+    // directly.
+    CSSRect visibleRectCSS = CSSPixel::FromAppUnits(mEffectsInfo.mVisibleRect);
+    return Some(visibleRectCSS * mPuppetWidget->GetDefaultScale());
+  }
+
+  Maybe<LayoutDeviceToLayoutDeviceMatrix4x4> inverse =
+      mChildToParentConversionMatrix->MaybeInverse();
+  if (!inverse) {
+    return Nothing();
+  }
+
+  // Convert the remote document visible rect to the coordinate system of the
+  // iframe document.
+  Maybe<LayoutDeviceRect> rect = UntransformBy(
+      *inverse,
+      ViewAs<LayoutDevicePixel>(
+          mTopLevelViewportVisibleRectInBrowserCoords,
+          PixelCastJustification::ContentProcessIsLayerInUiProcess),
+      LayoutDeviceRect::MaxIntRect());
+  if (!rect) {
+    return Nothing();
+  }
+
+  return rect;
+}
+
 ScreenIntRect BrowserChild::GetOuterRect() {
   LayoutDeviceIntRect outerRect =
       RoundedToInt(mUnscaledOuterRect * mPuppetWidget->GetDefaultScale());